internals improvements

vtjnash opened the pull request:
Some various improvements to some internals, observed while doing other work

vtjnash committed:
improve jl_pointer_egal for more cases

vtjnash committed:
codegen: fix needloop for replacefield

Wrong previously for String and SimpleVector, and also less optimal than was feasible

vtjnash committed:
use more efficient calling convention for calls with few arguments

vtjnash committed:
add more info to SIGUSR1 printing

This is similar to what the kernel prints for us when sent SIGINFO, and
can be sometimes useful for getting the pid of a process, and other
status details.

aviatesk commented on src/codegen.cpp:
```diff
@@ -1777,7 +1777,7 @@ static std::pair<bool, bool> uses_specsig(jl_method_instance_t *lam, jl_value_t
     if (prefer_specsig)
         return std::make_pair(true, false);
-    if (!deserves_retbox(rettype) && !jl_is_datatype_singleton((jl_datatype_t*)rettype))
+    if (!deserves_retbox(rettype) && !jl_is_datatype_singleton((jl_datatype_t*)rettype) && rettype != (jl_value_t*)jl_bool_type)
```
What's the rationale of this change ?

vtjnash replied:
This part estimates the calling convention cost, and there is no cost to a bool

vtjnash committed:
atomics: fix replacefield loop

We would previously loop back to the wrong point in the code, and cause
potentially chaos and malformed IR.

vtjnash committed:
FileWatching: fix typo on #42926 (2539a679ab)
